pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,539 @@
+use cfg_primitives::LP_DEFENSIVE_WEIGHT;
+use cfg_traits::liquidity_pools::{
+	InboundMessageHandler, LPEncoding, MessageQueue, MessageSender, Proof, RouterProvider,
+};
+use cfg_types::domain_address::{Domain, DomainAddress};
+use frame_support::{
+	dispatch::DispatchResult,
+	ensure,
+	pallet_prelude::{Decode, Encode, Get, TypeInfo},
+	weights::Weight,
+};
+use parity_scale_codec::MaxEncodedLen;
+use sp_arithmetic::traits::{EnsureAddAssign, EnsureSub, SaturatedConversion};
+use sp_runtime::DispatchError;
+
+use crate::{
+	message::GatewayMessage, Config, Error, Pallet, PendingInboundEntries, Routers, SessionIdStore,
+};
+
+/// Type that holds the information needed for inbound message entries.
+#[derive(Debug, Encode, Decode, Clone, Eq, MaxEncodedLen, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct MessageEntry<T: Config> {
+	/// The session ID for this entry.
+	pub session_id: T::SessionId,
+
+	/// The sender of the inbound message.
+	pub domain_address: DomainAddress,
+
+	/// The LP message.
+	pub message: T::Message,
+
+	/// The expected proof count for processing one or more of the provided
+	/// message.
+	///
+	/// NOTE - this gets increased by the `expected_proof_count` for a set of
+	/// routers (see `get_expected_proof_count`) every time a new identical
+	/// message is submitted.
+	pub expected_proof_count: u32,
```

Comment:
Do we really need this field? My rationale:

This information already exists indirectly in the `PendingInboundEntries`, and can be computed for a specific message on-the-fly in `process_inbound_message()`, checking the storages for the routers that message domain belong and counting them.

So maybe we do not need to store it, and later read it and update that field along the code. WDYT?

Reply:
> This information already exists indirectly in the `PendingInboundEntries`, and can be computed for a specific message on-the-fly in `process_inbound_message()`, by checking the storages for the routers to which the message domain belongs and counting them.

Can you help me and tell me which storage(s) in the router we would check for this?

IIUC, this would be at least one storage read per router right. Given that is is recommended to keep the weight of processing inbound messages low for `on_idle`, I currently support keeping this field here. Looking forward to understanding your path better @lemunozm!

Reply:
You anyway needs to look into all entries later to check if the message must be executed counting votes. Anyways the reads will be 2-3, the same as configured routers, so very very low.

Reply:
In `PendingInboundEntries`. If, for a message proof hash, you get a list with the second keys: the `router_ids`, and count them, you have the `expected_proof_count` value

Reply:
> Do we really need this field?
Yes, again required for ensuring proper processing of duplicate messages. For example, if we wouldn't do this for a case such as [this](https://github.com/centrifuge/centrifuge-chain/blob/88bdad7b58cbb3661c405711fba34a959b5589ae/pallets/liquidity-pools-gateway/src/tests.rs#L1817), the message would be "lost" after the first proof is received.

The name is a bit confusion tho, maybe we should rename it to `total_expected_proof_count`.

Reply:
I'm not sure if I fully understood what you said. Even if the first message is a proof, we still have the information about the required routers for that domain, isn't? So we can compute the count. 

But don't let this to block us from merging this 👍🏻

Reply:
Let's sort this out tomorrow and merge this beauty!

Reply:
We can even merge this, and solve this (if it's required) in a next up PR

Reply:
> Basically, when we get the second message, the expected count gets increased by the expected counts for processing a message for this case, which is one. This ensures that the message entry doesn't get removed from storage when the first proof is received.

Now I understand you Cosmin. I think you're right 👍🏻. Such a complex casuistics has this feature!

To resume, basically, if we receive two identical messages with no proofs, we're storing the same message twice in the same entry, so we anyways need some kind of counter to identify that state. We use the above counter, and we take advance of having a counter to also count the expected proofs 👍🏻